openwisp_notifications/static/openwisp-notifications/js/notifications.js

```diff
@@ -69,24 +74,51 @@ function initNotificationDropDown($) {
     $(document).focusin(function(e){
         // Hide notification widget if focus is shifted to an element outside it
         e.stopPropagation();
-        if ($('.ow-notification-dropdown').has(e.target).length === 0){
+        if (
+            $('.ow-notification-dropdown').has(e.target).length === 0 &&
+            // Do not hide if the user is interacting with the notification dialog
+            !$('.ow-dialog-overlay').is(':visible')
+        ) {
             // Don't hide if focus changes to notification bell icon
             if (e.target != $('#openwisp_notifications').get(0)) {
                 $('.ow-notification-dropdown').addClass('ow-hide');
             }
         }
     });
 
-    $('.ow-notification-dropdown').on('keyup', '*', function(e){
+    $(document).on('keyup', '*', function(e){
         e.stopPropagation();
+        if (e.keyCode !== 27) {
+            return;
+        }
         // Hide notification widget on "Escape" key
-        if (e.keyCode == 27){
+        if ($('.ow-dialog-overlay').is(':visible')) {
+            $('.ow-dialog-overlay').addClass('ow-hide');
+            $('.ow-message-target-redirect').addClass('ow-hide');
+        } else {
             $('.ow-notification-dropdown').addClass('ow-hide');
+        }
+        if ($(e.target).is($('.ow-notification-dropdown'))) {
             $('#openwisp_notifications').focus();
```

Comment:
Thinking out loud here: 

Shouldn't this be

```suggestion
            $('.ow-notification-dropdown').addClass('ow-hide');
            $('#openwisp_notifications').focus();
```

Why do we need to check `if ($(e.target).is($('.ow-notification-dropdown')))`